docs/architecture/autofill/index.md

```diff
@@ -0,0 +1,123 @@
+| ---------------------------- | --------------------------------------------------------------------------------------------------------------------------------------- |
+| `runtime.background.ts`      | Handles incoming requests related to core extension functionality.                                   |
+| `notification.background.ts` | Handles incoming requests related to the notification bar.                                  |
+| `commands.background.ts`     | Handles incoming requests related to keyboard commands (including autofill).                                    |
+| `contextMenu.background.ts`  | Handles context menu actions (including autofill).                                    |
+| `main.background.js`         | Bootstraps the extension. It is relevant here only because it (arbitrarily) contains the `collectPageDetailsForContentScript()` method. |
+
+In order to support browsers that will require the
+[Chrome Extension Manifest v3](https://developer.chrome.com/docs/extensions/mv3/intro/)
+specifications, the Bitwarden extension also handles the keyboard shortcut command through an
+`onCommandListener`. This listener is registered with `chrome.commands.onCommand.addListener`. The
+`onCommandListener` is responsible for initiating the `AutoFillActiveTabCommand` to trigger the
+Autofill functionality, similar to how the `commands.background.ts` background page handles incoming
+keyboard command messages.
+
+### Messaging
+
+We have established that the Bitwarden extension uses content scripts and background pages or
+listeners in order to perform Autofill. The last piece of the puzzle is the communication between
+them. The Autofill architecture leverages the
+[extension messaging API](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/sendMessage)
+in order to communicate between the extension UI, the background pages, and the content scripts that
+are running on each browser tab.
+
+```kroki type=plantuml
+@startuml
+
+ package "Browser Tab" as T {
+    [Page Source]
+    [Content Scripts]
+ }
+
+ node "Bitwarden Browser Extension" as E {
+    [Browser Extension UI]
+    [Background Pages]
+ }
+
+ [Background Pages] --> [Content Scripts] : tabSendMessage
+ [Content Scripts] --> [Background Pages] : sendMessage
+
+@enduml
+```
+
+#### Sending a request from the content script to the extension
+
+To send a request from a content script to the extension, we have two services provided:
+
+- When Dependency Injection is available, the `BrowserMessagingService` should be used, with its
+  provided `send()` method.
+- When no Dependency Injection is available, the static `BrowserApi` should be used, with its
+  `sendMessage()` method.
+
+On the extension, we attach with `chrome.runtime.onMessage.addListener` to receive the messages
```

Comment:
It looks like there's a wrapper method for this as well: `BrowserApi.messageListener`.

Reply:
Thank you!